lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1229 @@
+	mlir::rlc::ConstraintsLattice* currentLattice, mlir::rlc::ConstraintsAnalysis* analysis)
+{
+	// Aliases for more clarity
+	const auto& result=this->getOperation()->getResult(0);
+	const auto& op1=this->getOperation()->getOperand(0);
+	const auto& op2=this->getOperation()->getOperand(1);
+
+	// Do nothing if the lhs of the operation is not found in the lattice
+	if (not currentLattice->foundInRanges(std::make_pair(result,true)) and
+		not currentLattice->foundInRanges(std::make_pair(result,false)))
+		return false;
+
+	// TODO: add sanity check for operation type (i.e. do not work with floats)
+	
+	// Call the interface operation which updates the lattice accordingly
+	llvm::TypeSwitch<mlir::Operation*>(this->getOperation())
+	.Case<mlir::rlc::CA_CommutativeOp>([&](mlir::rlc::CA_CommutativeOp addop){
+		return currentLattice->findInterestingOperandAndUpdate(this->getOperation(),analysis,result);
+	});
+
+	return false;
+
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Operation Bindings : LEAVE THEM HERE OR ELSE NOTHING WILL WORK
+//////////////////////////////////////////////////////////////////////////////
+
+bool mlir::rlc::AddOp::constraintsAnalyze(
+	mlir::rlc::ConstraintsLattice* currentLattice, mlir::rlc::ConstraintsAnalysis* analysis)
+{
+	return mlir::dyn_cast<mlir::rlc::CA_CommutativeOp>(this->getOperation()).constraintsAnalyze(currentLattice,analysis);
+}
+
+bool mlir::rlc::MultOp::constraintsAnalyze(
+	mlir::rlc::ConstraintsLattice* currentLattice, mlir::rlc::ConstraintsAnalysis* analysis)
+{
+	return mlir::dyn_cast<mlir::rlc::CA_CommutativeOp>(this->getOperation()).constraintsAnalyze(currentLattice,analysis);
+}
+
+bool mlir::rlc::SubOp::constraintsAnalyze(
+	mlir::rlc::ConstraintsLattice* currentLattice, mlir::rlc::ConstraintsAnalysis* analysis)
+{
+	return mlir::dyn_cast<mlir::rlc::CA_NonCommutativeOp>(this->getOperation()).constraintsAnalyze(currentLattice,analysis);
+}
+
+bool mlir::rlc::DivOp::constraintsAnalyze(
+	mlir::rlc::ConstraintsLattice* currentLattice, mlir::rlc::ConstraintsAnalysis* analysis)
+{
+	return mlir::dyn_cast<mlir::rlc::CA_NonCommutativeOp>(this->getOperation()).constraintsAnalyze(currentLattice,analysis);
+}
+
+bool mlir::rlc::ReminderOp::constraintsAnalyze(
+	mlir::rlc::ConstraintsLattice* currentLattice, mlir::rlc::ConstraintsAnalysis* analysis)
+{
+	return mlir::dyn_cast<mlir::rlc::CA_NonCommutativeOp>(this->getOperation()).constraintsAnalyze(currentLattice,analysis);
+}
+
+bool mlir::rlc::MinusOp::constraintsAnalyze(
+	mlir::rlc::ConstraintsLattice* currentLattice, mlir::rlc::ConstraintsAnalysis* analysis)
+{
+	return mlir::dyn_cast<mlir::rlc::CA_NonCommutativeOp>(this->getOperation()).constraintsAnalyze(currentLattice,analysis);
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// NONCOMMUTATIVE OPERATION (TODO)
+//////////////////////////////////////////////////////////////////////////////
+
```

Comment:
if this does nothing, can't you not tag operations with it?

Reply:
Ok so that is what I would like to do but I did not find an easy way to do so.

Because, for ex. line 1342 of Operations.td (of this branch):

` def RLC_DivOp : RLC_InternalOp<"div", DeclareOpInterfaceMethods<CA_NonCommutativeOp>> { `

The division, which is noncommutative, derives from RLC_InternalOp (as all the arithmetic operations do).

Now the standard definition at line 1254 of the file (not in this branch, in the main) is 

` class RLC_InternalOp<string mnemonic> : RLC_Dialect<mnemonic, [TypeCheckable]> `

And I needed to add an interface to this class because if I tried to add it as above it would not compile.
So i modified it as

` class RLC_InternalOp<string mnemonic, OpInterface CA_Interface> : RLC_Dialect<mnemonic, [TypeCheckable, CA_Interface]> { `

But this requires that RLC_InternalOp has an interface, now if i do not add it to a definition it does not compile. I tried the solution with a list:

` class RLC_InternalOp<string mnemonic, list<OpInterface> CA_Interface=[]> : RLC_Dialect<mnemonic, [TypeCheckable, CA_Interface]> { `

But I did not find a way to correctly concatenate ` [TypeCheckable, CA_Interface]`, as one is a single interface and one is a list.

So for the moment I have to add the CA_NonCommutativeOp to all the operations